### PR TITLE
Applications v3

### PR DESCRIPTION
**Pull request by IvanBorislavovDimitrov**
LMCROSSITXSADEPLOY-1749

### CODE REVIEW

**Review by radoslav-d — APPROVED** [review]
Everything seems right

**Review by radoslav-d — APPROVED** [review]

File: multiapps-controller-process/src/main/java/module-info.java [hunk 1]
```diff
@@ -50,6 +50,8 @@
     requires spring.core;
     requires spring.security.oauth2;
     requires spring.web;
+    requires reactor.netty;
+    requires io.netty.handler;
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/module-info.java:54]: Why are these needed?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java [hunk 2]
```diff
@@ -94,6 +99,34 @@ public StepPhase executeAsyncStep(ProcessContext context) throws FileStorageExce
         return StepPhase.DONE;
     }
 
+    private CloudPackage createDockerPackage(CloudControllerClient client, CloudApplicationExtended application) {
+        UUID applicationGuid = client.getApplicationGuid(application.getName());
+        return client.createDockerPackage(applicationGuid, application.getDockerInfo());
+    }
+
+    private String getNewApplicationDigest(ProcessContext context, String appArchiveId, String fileName) throws FileStorageException {
+        return fileService.processFileContent(context.getVariable(Variables.SPACE_GUID), appArchiveId,
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java:108]: Can `appArchiveId` be retrieved in this method instead of passing it as a parameter?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java [hunk 3]
```diff
@@ -63,17 +64,21 @@ public StepPhase executeAsyncStep(ProcessContext context) throws FileStorageExce
 
         MtaArchiveElements mtaArchiveElements = context.getVariable(Variables.MTA_ARCHIVE_ELEMENTS);
         String moduleFileName = mtaArchiveElements.getModuleFileName(applicationToProcess.getModuleName());
+        CloudControllerClient client = context.getControllerClient();
         if (moduleFileName == null) {
             getStepLogger().debug(Messages.NO_CONTENT_TO_UPLOAD);
+            if (applicationToProcess.getDockerInfo() != null) {
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java:70]: If I understand correctly, this flow means that there isn't a file entry in the MTAR associated with this module because it is described as a docker application, is this right?
If so, how did we operate with docker apps so far?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java [hunk 4]
```diff
@@ -94,6 +99,34 @@ public StepPhase executeAsyncStep(ProcessContext context) throws FileStorageExce
         return StepPhase.DONE;
     }
 
+    private CloudPackage createDockerPackage(CloudControllerClient client, CloudApplicationExtended application) {
+        UUID applicationGuid = client.getApplicationGuid(application.getName());
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java:103]: Why isn't the guid in the `CloudApplicationExtended` object used but a request to the CC is made to fetch the guid?

File: multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/steps/CreateOrUpdateAppStepTest.java [hunk 5]
```diff
@@ -92,16 +92,19 @@ void testCreateApplicationFromDockerImage() {
                                                                                           .password("somePassword")
                                                                                           .build())
                                                    .build();
+        Staging staging = ImmutableStaging.builder()
```
Comment (radito3) [inline multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/steps/CreateOrUpdateAppStepTest.java:95]: Maybe `dockerStaging` would be more descriptive

File: multiapps-controller-process/src/main/java/module-info.java (hunk 1 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/module-info.java:54]: Error occurred during initialization of boot layer
java.lang.module.FindException: Module reactor.netty not found, required by com.sap.cloudfoundry.client.facade
They are required as far as I understood

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java (hunk 2 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java]: Yes, it can.

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java (hunk 3 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java:70]: In v2 the docker apps do not require packages, they are just created as docker image apps and started. In v3 it is required to create a docker package to deploy the application. If there is no module name the application is probably a docker app but I added this check just in case if the user does not provide docker parameters otherwise a null pointer will occur.

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java (hunk 4 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java:102]: Because it set in BuildApplicationDeployModelStep and its guid is unknown -> the id is null.

File: multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/steps/CreateOrUpdateAppStepTest.java (hunk 5 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/steps/CreateOrUpdateAppStepTest.java]: Done

**Review by radito3 — APPROVED** [review]
